libretro-rs/src/environment.rs

```diff
@@ -0,0 +1,169 @@
+}
+
+pub trait SetEnvironmentEnvironment: RetroEnvironment {
+  fn set_support_no_game(&mut self, data: bool) -> bool {
+    unsafe { self.set_raw(RETRO_ENVIRONMENT_SET_SUPPORT_NO_GAME, &data) }
+  }
+}
+impl <T> SetEnvironmentEnvironment for T where T: RetroEnvironment {}
+
+pub trait InitEnvironment: RetroEnvironment {}
+impl <T> InitEnvironment for T where T: RetroEnvironment {}
+
+pub trait SetPortDeviceEnvironment: RetroEnvironment {}
+impl <T> SetPortDeviceEnvironment for T where T: RetroEnvironment {}
+
+pub trait ResetEnvironment: RetroEnvironment {}
+impl <T> ResetEnvironment for T where T: RetroEnvironment {}
+
+pub trait RunEnvironment: RetroEnvironment {
+  /// Requests that the frontend shut down. The frontend can refuse to do this, and return false.
+  fn shutdown(&mut self) -> bool {
+    unsafe { self.cmd_raw(RETRO_ENVIRONMENT_SHUTDOWN) }
+  }
+
+  fn set_geometry(&mut self, geometry: RetroGameGeometry) -> bool {
+    let data: retro_game_geometry = geometry.into();
+    unsafe { self.set_raw(RETRO_ENVIRONMENT_SET_GEOMETRY, &data) }
+  }
+}
+impl <T> RunEnvironment for T where T: RetroEnvironment {}
+
+pub trait SerializeSizeEnvironment: RetroEnvironment {}
+impl <T> SerializeSizeEnvironment for T where T: RetroEnvironment {}
+
+pub trait SerializeEnvironment: RetroEnvironment {}
+impl <T> SerializeEnvironment for T where T: RetroEnvironment {}
+
+pub trait UnserializeEnvironment: RetroEnvironment {}
+impl <T> UnserializeEnvironment for T where T: RetroEnvironment {}
+
+pub trait CheatResetEnvironment: RetroEnvironment {}
+impl <T> CheatResetEnvironment for T where T: RetroEnvironment {}
+
+pub trait CheatSetEnvironment: RetroEnvironment {}
+impl <T> CheatSetEnvironment for T where T: RetroEnvironment {}
+
+pub trait LoadGameEnvironment: RetroEnvironment {
+  fn set_pixel_format(&mut self, format: RetroPixelFormat) -> bool {
+    GetSystemAvInfoEnvironment::set_pixel_format(self, format)
+  }
+}
+impl <T> LoadGameEnvironment for T where T: RetroEnvironment {}
+
+pub trait GetSystemAvInfoEnvironment: RetroEnvironment {
+  fn set_pixel_format(&mut self, format: RetroPixelFormat) -> bool {
+    unsafe { self.set_raw(RETRO_ENVIRONMENT_SET_PIXEL_FORMAT, &u32::from(format)) }
+  }
+}
+impl <T> GetSystemAvInfoEnvironment for T where T: RetroEnvironment {}
+
+pub trait GetRegionEnvironment: RetroEnvironment {}
+impl <T> GetRegionEnvironment for T where T: RetroEnvironment {}
+
+pub trait LoadGameSpecialEnvironment: RetroEnvironment {}
+impl <T> LoadGameSpecialEnvironment for T where T: RetroEnvironment {}
+
+pub trait UnloadGameEnvironment: RetroEnvironment {}
+impl <T> UnloadGameEnvironment for T where T: RetroEnvironment {}
+
+pub trait GetMemoryDataEnvironment: RetroEnvironment {}
+impl <T> GetMemoryDataEnvironment for T where T: RetroEnvironment {}
+
+pub trait GetMemorySizeEnvironment: RetroEnvironment {}
+impl <T> GetMemorySizeEnvironment for T where T: RetroEnvironment {}
```

Comment:
I assume these traits are a first step towards #2? I have some thoughts on how to do this that differ from this implementation. Let's make those changes in a separate PR, so we can consider them in isolation.